apps/dashboard/src/utils/schema.ts

```diff
@@ -0,0 +1,167 @@
+  requiredFields: Readonly<Array<string>>;
+}) => {
+  const isRequired = requiredFields.includes(key);
+
+  let stringValue:
+    | z.ZodString
+    | z.ZodEffects<z.ZodTypeAny>
+    | z.ZodEnum<[string, ...string[]]>
+    | z.ZodDefault<z.ZodTypeAny> = z.string();
+
+  if (format) {
+    stringValue = handleStringFormat({
+      value: stringValue,
+      key,
+      format,
+    });
+  } else if (pattern) {
+    stringValue = handleStringPattern({
+      value: stringValue,
+      key,
+      pattern,
+    });
+  } else if (enumValues) {
+    stringValue = handleStringEnum({
+      key,
+      enumValues: enumValues as [string, ...string[]],
+    });
+  } else if (isRequired) {
+    stringValue = stringValue.min(1, `${capitalize(key)} is required`);
+  }
+
+  if (defaultValue) {
+    stringValue = stringValue.default(defaultValue as string);
+  }
+
+  // remove empty strings
+  return stringValue.transform((val) => (val === '' ? undefined : val));
+};
+
+/**
+ * Transform JSONSchema to Zod schema.
+ * The function will recursively build the schema based on the JSONSchema object.
+ * It removes empty strings and objects with empty required fields during the transformation phase after parsing.
+ */
+export const buildDynamicZodSchema = (obj: JSONSchema): z.AnyZodObject => {
+  const properties = typeof obj === 'object' ? (obj.properties ?? {}) : {};
+  const requiredFields = typeof obj === 'object' ? (obj.required ?? []) : [];
+
+  const keys: Record<string, z.ZodTypeAny> = Object.keys(properties).reduce((acc, key) => {
+    const jsonSchemaProp = properties[key];
+    if (typeof jsonSchemaProp !== 'object') {
+      return acc;
+    }
+
+    let zodValue: ZodValue;
+    const { type, format, pattern, enum: enumValues, default: defaultValue, required } = jsonSchemaProp;
+    const isRequired = requiredFields.includes(key);
+
+    if (type === 'object') {
+      zodValue = buildDynamicZodSchema(jsonSchemaProp);
+      if (defaultValue) {
+        zodValue = zodValue.default(defaultValue);
+      }
+      zodValue = zodValue.transform((val) => {
+        const hasAnyRequiredEmpty = required?.some((field) => val[field] === '' || val[field] === undefined);
+        // remove object if any required field is empty or undefined
+        return hasAnyRequiredEmpty ? undefined : val;
+      });
+    } else if (type === 'string') {
+      zodValue = handleStringType({ key, requiredFields, format, pattern, enumValues, defaultValue });
+    } else {
+      zodValue = z.number(isRequired ? { message: `${capitalize(key)} is required` } : undefined);
+      if (defaultValue) {
+        zodValue = zodValue.default(defaultValue as number);
+      }
+    }
+
+    if (!isRequired) {
+      zodValue = zodValue.optional() as ZodValue;
+    }
+
+    return { ...acc, [key]: zodValue };
+  }, {});
+
+  return z.object({ ...keys });
+};
+
+/**
+ * Build default values based on the JSONSchema object.
+ */
+export const buildDefaultValues = (obj: JSONSchema): object => {
```

Comment:
This util takes the `dataSchema: JSONSchema` and retrieves the default values from it.

Reply:
@SokratisVidros the work done here shows how unnecessary was using the json schema if we need to manually parse it in the FE, we should have made the zod available via shared library and convert the zod to jsonschema for the library usage

Reply:
I am not sure I understand the comment. If we hardcode the values per step how can we support dynamic configurations?